### ISSUE TITLE
UnloadRecord fails to cleanup hasMany relationships and causes computed properties to explode

### ISSUE THREAD

**Issue by Kilowhisky** [emberjs/data#5524]
This is related to all the other UnloadRecord bugs floating around (including the ones i've opened). This one in particular is present in 3.2 and has to do with unloading a record that is being watched through a hasMany relationship by a computed property. 

What happens is that when records are unloaded the hasMany array contains null / undefined entries where the unloaded record previously resided. This causes anything watching the array to explode if it doesn't have null/undefined checking which happens to be all the computed properties ember offers.

First of all here is a twiddle showing the error:
https://ember-twiddle.com/5bee61b69a3813524b1d83b319c4fc30?openFiles=routes.application.js%2C

Here is the timeline for what is happening:

1) A main record is loaded (in my case this is an "asset"). It contains a hasMany relationship to a model called "position" and also contains a computed property that sorts the "positions".
2) The "positions" are loaded on demand and reference the "asset" that they belongTo. 
3) The UI shows the sorted positions via Asset > PositionsSorted
4) Some of the "positions" are then unloaded on the fly.
5) The hasMany still contains null / undefined references where the records used to be in the array. 
6) The computed property for sorting explodes because it can't sort undefined / null records.

```
"Error: Assertion Failed: Cannot call get with 'utc' on an undefined object.
    at new EmberError (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:14719:25)
    at Object.assert (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:14270:15)
    at get (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:26810:64)
    at https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:37177:64
    at Array.sort (native)
    at sortByNormalizedSortProperties (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:37171:40)
    at Class._emberMetal.ComputedProperty.dependentKeys (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:37149:16)
    at ComputedProperty.get (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:27503:30)
    at get (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:26832:27)
    at NestedPropertyReference.compute (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:16067:44)"
```

**Comment by runspired** [comment]
Given the use of an array observer on an async relationship, I'm unsure this is a bug vs just poor design. I'll need to consider whether we intend to support this situation but the answer is likely that we don't.

**Comment by runspired** [comment]
@sly7-7 @Kilowhisky does this issue still exist today? We've fixed so many things around this behavior I'd imagine that if it wasn't fixed by 3.27 it's at least fixed by 3.29-canary. Would love a check!
